drake/solvers/test/testOptimizationProblem.cpp

```diff
@@ -58,7 +60,7 @@ class SixHumpCamelObjective : public TemplatedDifferentiableFunction<SixHumpCame
 void sixHumpCamel() {
   OptimizationProblem prog;
   auto x = prog.addContinuousVariables(2);
-  auto objective = prog.addCost(make_shared<SixHumpCamelObjective>());
+  auto objective = prog.addCost(SixHumpCamelObjective());
```

Comment:
This is how type erasure works.  The `enable_if`-ed member gets instantiated with `SixHumpCamelObjective` as its `F` type.  Then an instantiation of `ConstraintImpl<SixHumpCamelObjective>` is generated and an instance is allocated with `make_shared`.  It stores a copy of SixHumpCamelObjective by (const) value.  It is accessed via virtual dispatch in the `Constraint` hierarchy because `ConstraintImpl<F>` implements the `eval` methods by calling (statically) into the `F` type (through `FunctionTraits<F>` lookups to be more general).

If needed this technique can be tweaked to also allow passing a function structure by reference or shared_ptr.  I'll look at this in the next revision too.